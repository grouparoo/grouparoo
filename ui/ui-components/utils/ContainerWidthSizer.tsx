import { ReactNode, Component } from "react";

interface State {
  width: number;
  height: number;
}

interface Props {
  children: (state: State) => ReactNode;
}

export class ContainerWidthSizer extends Component<Props, State> {
  state = {
    width: 0,
    height: 0,
  };

  container: HTMLDivElement;

  componentDidMount() {
    const container = this.container.getBoundingClientRect();
    const width = Math.max(100, container.width);
    const height = Math.max(200, container.height);
    this.setState({ width, height });
  }

  render() {
    return (
      <div
        style={{ width: "100%", height: "100%" }}
        ref={(c) => (this.container = c)}
      >
        {this.props.children(this.state)}
      </div>
    );
  }
}
