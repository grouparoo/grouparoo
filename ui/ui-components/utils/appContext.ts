import { GetServerSidePropsContext, NextPageContext } from "next";
import { AppContext } from "next/app";

function isAppContext(ctx: unknown): ctx is AppContext {
  return !!ctx.hasOwnProperty("ctx");
}
export const getRequestContext =
  (ctx: AppContext | NextPageContext | GetServerSidePropsContext) => () => ({
    req: isAppContext(ctx) ? ctx.ctx.req : ctx.req,
    res: isAppContext(ctx) ? ctx.ctx.res : ctx.res,
  });
