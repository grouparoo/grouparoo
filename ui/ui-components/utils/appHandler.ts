import { Models } from "./apiData";
import { EventDispatcher } from "./eventDispatcher";

export class AppHandler extends EventDispatcher<Models.AppType> {}
