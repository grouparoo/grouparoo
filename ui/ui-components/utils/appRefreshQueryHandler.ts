import { Models } from "./apiData";
import { EventDispatcher } from "./eventDispatcher";

export class AppRefreshQueryHandler extends EventDispatcher<Models.AppType> {}
