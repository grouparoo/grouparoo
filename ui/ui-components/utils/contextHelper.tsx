export type ContextValue<T = unknown> = [Context: React.Context<T>, value: T];

// Note: Once strict mode is on, we can more strongly type this
export const renderContextProvider = <T extends any>(
  Context: React.Context<T>,
  value: T,
  children: React.ReactNode
) => <Context.Provider value={value}>{children}</Context.Provider>;

export const renderNestedContextProviders = <
  TContextValues extends ContextValue<T>[],
  T extends object = object
>(
  contextValues: TContextValues,
  children: React.ReactNode
): JSX.Element => {
  {
    if (contextValues.length) {
      const [[Context, value]] = contextValues;
      return renderContextProvider(
        Context,
        value,
        renderNestedContextProviders(contextValues.slice(1), children)
      );
    }

    return <>{children}</>;
  }
};
