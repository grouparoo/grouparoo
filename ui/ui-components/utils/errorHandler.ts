import { EventDispatcher } from "./eventDispatcher";

export class ErrorHandler extends EventDispatcher<{ error: string | Error }> {
  error: { error: string | Error };

  constructor() {
    super();

    this.error = null;

    this.subscribe("_internal", (e) => {
      this.error = e;
    });
  }
}
