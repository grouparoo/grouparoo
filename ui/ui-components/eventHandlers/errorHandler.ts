import { EventDispatcher } from "../utils/eventDispatcher";

export class ErrorHandler extends EventDispatcher<{ message: string | Error }> {
  message: string | Error;

  constructor() {
    super();

    this.message = null;

    this.subscribe("_internal", (e) => {
      this.message = e.message;
    });
  }
}
