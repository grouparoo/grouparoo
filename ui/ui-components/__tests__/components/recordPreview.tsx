import RecordPreview from "../../components/property/recordPreview";
import { ErrorHandler } from "../../utils/errorHandler";
import { renderAndWait } from "../__utils__/renderAndWait";

jest.mock("next/router", () => {
  const nextRouter = jest.requireActual("next/router");
  nextRouter.useRouter = jest.fn(() => ({
    route: "/",
    query: {
      recordId: "asdf",
    },
    replace: jest.fn(),
    push: jest.fn(),
  }));
  return nextRouter;
});

const execApi = jest.fn(() => {
  return Promise.resolve({});
});
jest.mock("../../hooks/useApi", () => ({
  useApi: () => ({
    execApi,
  }),
}));

const renderDefaultRecord = (overrideProps: any = {}) =>
  renderAndWait(
    <RecordPreview
      errorHandler={new ErrorHandler()}
      localFilters={{}}
      property={{
        id: "test-id",
        key: "test-key",
        type: "test-type",
        options: {},
      }}
      {...overrideProps}
    />
  );

describe("property / recordPreview", () => {
  const dateMock = jest
    .spyOn(Date.prototype, "toLocaleString")
    .mockReturnValue("9/8/2021, 6:58:38 PM");

  afterAll(() => {
    jest.unmock("next/router");
    jest.unmock("../../hooks/useApi");
    dateMock.mockRestore();
  });
  afterEach(() => {
    execApi.mockReset();
  });

  it("should render with no record preview", async () => {
    const record = await renderDefaultRecord();
    expect(record.container).toMatchSnapshot();
  });

  it("should render", async () => {
    execApi.mockImplementation(() => {
      return Promise.resolve({
        record: {
          id: "asdf",
          properties: {
            date: { type: "date", values: [1631127518020] },
          },
        },
      });
    });

    const record = await renderDefaultRecord();
    expect(record.container).toMatchSnapshot();
  });

  it("should render with null date", async () => {
    execApi.mockImplementation(() => {
      return Promise.resolve({
        record: {
          id: "asdf",
          properties: {
            date: { type: "date", values: [null] },
          },
        },
      });
    });

    const record = await renderDefaultRecord();
    expect(record.container).toMatchSnapshot();
  });
});
