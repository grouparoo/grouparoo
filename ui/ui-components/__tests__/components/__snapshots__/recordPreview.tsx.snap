// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`property / recordPreview should not render with a boolean date 1`] = `
<div>
  <div
    class="card bg-light"
  >
    <div
      class="card-body"
      style="text-align: center; color: white;"
    >
      <div
        class="card-title h5"
      >
        Example Record
      </div>
      <img
        class="rounded"
        src="https://www.gravatar.com/avatar/d41d8cd98f00b204e9800998ecf8427e?s=100.png"
        style="max-width: 100px; width: 100%;"
      />
      <br />
      <a
        class="card-link"
        href="/record/asdf/edit"
        style="color: white;"
      >
        View Record
      </a>
      <br />
      <small>
        change
      </small>
    </div>
    <div
      class="list-group list-group-flush"
    >
      <div
        class="list-group-item list-group-item-success"
      >
        <strong>
          test-key
        </strong>
        :
         
        ...
      </div>
      <div
        class="list-group-item list-group-item-light"
      >
        <strong>
          date
        </strong>
        :
         
      </div>
    </div>
  </div>
</div>
`;

exports[`property / recordPreview should not render with a null date 1`] = `
<div>
  <div
    class="card bg-light"
  >
    <div
      class="card-body"
      style="text-align: center; color: white;"
    >
      <div
        class="card-title h5"
      >
        Example Record
      </div>
      <img
        class="rounded"
        src="https://www.gravatar.com/avatar/d41d8cd98f00b204e9800998ecf8427e?s=100.png"
        style="max-width: 100px; width: 100%;"
      />
      <br />
      <a
        class="card-link"
        href="/record/asdf/edit"
        style="color: white;"
      >
        View Record
      </a>
      <br />
      <small>
        change
      </small>
    </div>
    <div
      class="list-group list-group-flush"
    >
      <div
        class="list-group-item list-group-item-success"
      >
        <strong>
          test-key
        </strong>
        :
         
        ...
      </div>
      <div
        class="list-group-item list-group-item-light"
      >
        <strong>
          date
        </strong>
        :
         
      </div>
    </div>
  </div>
</div>
`;

exports[`property / recordPreview should render a whole user profile including the date 1`] = `
<div>
  <div
    class="card bg-light"
  >
    <div
      class="card-body"
      style="text-align: center; color: white;"
    >
      <div
        class="card-title h5"
      >
        Example Record
      </div>
      <img
        class="rounded"
        src="https://www.gravatar.com/avatar/6c6821dfc7df35b8eb1c91d33707b690?s=100.png"
        style="max-width: 100px; width: 100%;"
      />
      <br />
      <a
        class="card-link"
        href="/record/rec_87cc054b-7cc2-4eb4-8932-db1011056517/edit"
        style="color: white;"
      >
        View Record
      </a>
      <br />
      <small>
        change
      </small>
    </div>
    <div
      class="list-group list-group-flush"
    >
      <div
        class="list-group-item list-group-item-success"
      >
        <strong>
          Draft Record Property
        </strong>
        :
         
        9/8/2021, 6:58:38 PM
      </div>
      <div
        class="list-group-item list-group-item-light"
      >
        <strong>
          userId
        </strong>
        :
         
        50
      </div>
      <div
        class="list-group-item list-group-item-light"
      >
        <strong>
          email
        </strong>
        :
         
        spenburton1d@demo.com
      </div>
      <div
        class="list-group-item list-group-item-light"
      >
        <strong>
          Deactivated
        </strong>
        :
         
        false
      </div>
      <div
        class="list-group-item list-group-item-light"
      >
        <strong>
          firstName
        </strong>
        :
         
        Sibeal
      </div>
      <div
        class="list-group-item list-group-item-light"
      >
        <strong>
          lastName
        </strong>
        :
         
        Penburton
      </div>
      <div
        class="list-group-item list-group-item-light"
      >
        <strong>
          LTV
        </strong>
        :
         
        341.66
      </div>
      <div
        class="list-group-item list-group-item-light"
      >
        <strong>
          lastPurchaseCategory
        </strong>
        :
         
        Toys
      </div>
      <div
        class="list-group-item list-group-item-light"
      >
        <strong>
          purchaseCount
        </strong>
        :
         
        6
      </div>
      <div
        class="list-group-item list-group-item-light"
      >
        <strong>
          fullName
        </strong>
        :
         
        Sibeal Penburton
      </div>
      <div
        class="list-group-item list-group-item-light"
      >
        <strong>
          Language
        </strong>
        :
         
        French
      </div>
      <div
        class="list-group-item list-group-item-light"
      >
        <strong>
          test_updated_at
        </strong>
        :
         
        9/8/2021, 6:58:38 PM
      </div>
    </div>
  </div>
</div>
`;

exports[`property / recordPreview should render successfully with no record preview 1`] = `
<div>
  <div
    class="card bg-light"
  >
    <div
      class="card-body"
      style="text-align: center;"
    >
      Record preview unavailable
    </div>
  </div>
</div>
`;

exports[`property / recordPreview should render with a Date date 1`] = `
<div>
  <div
    class="card bg-light"
  >
    <div
      class="card-body"
      style="text-align: center; color: white;"
    >
      <div
        class="card-title h5"
      >
        Example Record
      </div>
      <img
        class="rounded"
        src="https://www.gravatar.com/avatar/d41d8cd98f00b204e9800998ecf8427e?s=100.png"
        style="max-width: 100px; width: 100%;"
      />
      <br />
      <a
        class="card-link"
        href="/record/asdf/edit"
        style="color: white;"
      >
        View Record
      </a>
      <br />
      <small>
        change
      </small>
    </div>
    <div
      class="list-group list-group-flush"
    >
      <div
        class="list-group-item list-group-item-success"
      >
        <strong>
          test-key
        </strong>
        :
         
        ...
      </div>
      <div
        class="list-group-item list-group-item-light"
      >
        <strong>
          date
        </strong>
        :
         
        9/8/2021, 6:58:38 PM
      </div>
    </div>
  </div>
</div>
`;

exports[`property / recordPreview should render with a number date 1`] = `
<div>
  <div
    class="card bg-light"
  >
    <div
      class="card-body"
      style="text-align: center; color: white;"
    >
      <div
        class="card-title h5"
      >
        Example Record
      </div>
      <img
        class="rounded"
        src="https://www.gravatar.com/avatar/d41d8cd98f00b204e9800998ecf8427e?s=100.png"
        style="max-width: 100px; width: 100%;"
      />
      <br />
      <a
        class="card-link"
        href="/record/asdf/edit"
        style="color: white;"
      >
        View Record
      </a>
      <br />
      <small>
        change
      </small>
    </div>
    <div
      class="list-group list-group-flush"
    >
      <div
        class="list-group-item list-group-item-success"
      >
        <strong>
          test-key
        </strong>
        :
         
        ...
      </div>
      <div
        class="list-group-item list-group-item-light"
      >
        <strong>
          date
        </strong>
        :
         
        9/8/2021, 6:58:38 PM
      </div>
    </div>
  </div>
</div>
`;

exports[`property / recordPreview should render with a string date 1`] = `
<div>
  <div
    class="card bg-light"
  >
    <div
      class="card-body"
      style="text-align: center; color: white;"
    >
      <div
        class="card-title h5"
      >
        Example Record
      </div>
      <img
        class="rounded"
        src="https://www.gravatar.com/avatar/d41d8cd98f00b204e9800998ecf8427e?s=100.png"
        style="max-width: 100px; width: 100%;"
      />
      <br />
      <a
        class="card-link"
        href="/record/asdf/edit"
        style="color: white;"
      >
        View Record
      </a>
      <br />
      <small>
        change
      </small>
    </div>
    <div
      class="list-group list-group-flush"
    >
      <div
        class="list-group-item list-group-item-success"
      >
        <strong>
          test-key
        </strong>
        :
         
        ...
      </div>
      <div
        class="list-group-item list-group-item-light"
      >
        <strong>
          date
        </strong>
        :
         
        9/8/2021, 6:58:38 PM
      </div>
    </div>
  </div>
</div>
`;
